src/query/expression/src/evaluator.rs

```diff
@@ -2261,6 +2286,135 @@ impl<'a, Index: ColumnIndex> ConstantFolder<'a, Index> {
+            return None;
+        }
+
+        let (inner, dest_type, right) = match (&args[0], &args[1]) {
+            (
+                cast @ Expr::Cast {
+                    is_try: false,
+                    box expr,
+                    dest_type,
+                    ..
+                },
+                right,
+            )
+            | (
+                right,
+                cast @ Expr::Cast {
+                    is_try: false,
+                    box expr,
+                    dest_type,
+                    ..
+                },
+            ) if !cast.column_refs().is_empty() && right.column_refs().is_empty() => {
+                Some((expr, dest_type, right))
+            }
+            (
+                cast @ Expr::FunctionCall {
+                    function: cast_func,
+                    args,
+                    ..
+                },
+                right,
+            )
+            | (
+                right,
+                cast @ Expr::FunctionCall {
+                    function: cast_func,
+                    args,
+                    ..
+                },
+            ) if cast_func.signature.name.starts_with("to_")
+                && args.len() == 1
+                && !cast.column_refs().is_empty()
+                && right.column_refs().is_empty() =>
+            {
+                Some((&args[0], &cast_func.signature.return_type, right))
+            }
+            _ => None,
+        }?;
+
+        let (inner_expr, inner_domain) = self.fold_once(inner);
+        let inner_domain = inner_domain?;
+        let src_type = inner_expr.data_type();
+        if !Self::is_injective_cast(src_type, &inner_domain, dest_type) {
+            return None;
+        }
+        self.calculate_cast(*span, src_type, dest_type, &inner_domain)?; // check MayThrow
+
+        let (right, right_domain) = self.fold_once(right);
+        let cast_right_domain = self.calculate_cast(*span, dest_type, src_type, &right_domain?)?;
+
+        let cast = Expr::Cast {
+            span: None,
+            is_try: false,
+            expr: Box::new(right),
+            dest_type: src_type.clone(),
+        };
+        let func_expr =
+            check_function(*span, "eq", &[], &[inner_expr, cast], self.fn_registry).ok()?;
+
+        let calc_domain = func_expr
+            .as_function_call()
+            .unwrap()
+            .2
+            .eval
+            .as_scalar()
+            .unwrap()
+            .0;
+
+        let domain = match calc_domain(self.func_ctx, &[inner_domain, cast_right_domain]) {
+            FunctionDomain::MayThrow => None,
+            FunctionDomain::Full => None,
+            FunctionDomain::Domain(domain) => Some(domain),
+        };
+
+        Some((func_expr, domain))
+    }
+
+    fn is_injective_cast(src: &DataType, _src_domain: &Domain, dest: &DataType) -> bool {
+        if src == dest {
+            return true;
+        }
+
+        // Ignore FunctionDomain::MayThrow
```

Comment:
I did not see any codes to match the comment.

Reply:
This comment means that Ignore FunctionDomain::MayThrow is a precondition for is_injective_cast. It will be refined later.